Add FID metric

kamahori opened the pull request:
Fixes #998

Description:
Added FID metric which is used for GANs.
Most codes are incorporated from [here](https://github.com/clovaai/stargan-v2/blob/master/metrics/fid.py), so further optimization might be needed.

Check list:
* [x] New tests are added (if a new feature is added)
* [ ] New doc strings: description and/or example code are in RST format
* [ ] Documentation is updated (if required)

vfdev-5 commented:
@kamahori thanks for working on this metric ! I agree that it is not a simplest metric the intergrate.
I have a few comments about the API I was thinking about. Idea is that user can attach the metric to evaluator engine:

```python

fid_metric = FID()
evaluator = ...
fid_metric.attach(evaluator, "fid")

evaluator.run(validation_dataloader)
```
See here for more details: https://pytorch.org/ignite/concepts.html

Such that we can avoid dataloader setup from path etc. ImageNet pretrained model like Inception can an optional argument:
```python
class FID(Metric):
    def __init__(self, ..., test_model=None):
        ...
        if test_model is None:
            try:
                 from torchvision ... import inception
                 test_model = ..
            except ImportError:
                 ...
``` 
Maybe something else... I'll comment later.

Other remarks, probable, we should not use numpy or tqdm in the code. We can replace numpy by torch and tqdm can be attached by user on the evaluator.

kamahori commented:
Thanks for reviewing @vfdev-5 
I fixed several points.

> Other remarks, probable, we should not use numpy or tqdm in the code. We can replace numpy by torch and tqdm can be attached by user on the evaluator.

How about scipy? Replacing `scipy.linalg.sqrtm` with torch seems a bit complicated.

vfdev-5 commented:
> How about scipy? Replacing scipy.linalg.sqrtm with torch seems a bit complicated.

Good point, let me check if it can be possible or not at all...

kamahori commented:
Maybe we can use [this](https://github.com/steveli/pytorch-sqrtm) ?

vfdev-5 commented:
Thanks for the link. Seems like it is a wrap over scipy and backprop enabled. Backprop is not a requirement for us. Can you check if there is a FR on pytorch for that ? I wonder if it is difficult to implement ...

kamahori commented:
I also found [this](https://github.com/pytorch/pytorch/issues/25481) and [this](https://github.com/pytorch/pytorch/issues/9983), but it seems they are not exactly what we are looking for...
And FYI [here](https://github.com/scipy/scipy/blob/v1.5.2/scipy/linalg/_matfuncs_sqrtm.py#L117-L194) is the original scipy implementation.

vfdev-5 commented on ignite/metrics/fid.py:
```diff
@@ -0,0 +1,131 @@
+
+__all__ = ["FID"]
+
+class DefaultDataset(data.Dataset):
```
I think we should not deal with datasets, but have rely on batches provided by the evaluator.

vfdev-5 commented on ignite/metrics/fid.py:
```diff
@@ -0,0 +1,131 @@
+    def __init__(
+        self, 
+        output_transform: Callable = lambda x: x, 
+        input_path: os.path, 
```
we need not to have input_path and output_path as argument

vfdev-5 commented on ignite/metrics/fid.py:
```diff
@@ -0,0 +1,131 @@
+        if test_model is None:
+            try:
+                from torchvision import models
+                test_model = models.inception_v3()
```
Should not the model be pretrained=True ?

vfdev-5 commented on ignite/metrics/fid.py:
```diff
@@ -0,0 +1,131 @@
+        self._value = None
+
+    @reinit__is_reduced
+    def update(self, output) -> None:
```
We have to implement the metric such that it is computed in online manner:
- on update we accumulate some internal things (e.g. activations here)
- on compute we calculate the final metric using internal variables. (e.g compute frechet distance on mu and cov)

Let's define that the output is `y_pred` = batch of generated images = `generator(batch_x)` whatever the x is (scalar/vector for dcgans, image for cyclegans etc).
During the udpate we should think what to accumulate: activations average and covariance (not sure yet)

vfdev-5 commented on ignite/metrics/fid.py:
```diff
@@ -0,0 +1,131 @@
+        output_transform: Callable = lambda x: x, 
+        input_path: os.path, 
+        output_path: os.path,
+        test_model = None,
```
I like the implementation of fid here : https://github.com/Tetratrio/stylegan2_pytorch/blob/master/stylegan2/metrics/fid.py
and I think it would be better to call test_model as fid_model ?

vfdev-5 reviewed (commented):
@kamahori thanks for working on this metric !
I left some comments on how we need to reimplement the metric...
Honestly, it is not a simple metric to implement and hope we can find some nice solutions :)

vfdev-5 commented:
@kamahori if there is no "simple" option to make it without scipy, let's put the metric into `ignite.contrib.metrics` and use scipy. In this module, there are other metrics using other deps like sklearn etc

kamahori commented:
@vfdev-5 
Thank you for your comments. I fixed codes again.
I moved FID metric to `ignite.contrib.metrics` because I could not find any simple solution to scipy's problem.

vfdev-5 commented:
@kamahori thanks for the update ! It looks better now 👍  However, it remains a tough part about "online" fashion metric  computation, make it work with DDP and write tests

Let's handle them one by one.
1) Online fashion metric computation

Let me give you an example how it would work for a simple metric like accuracy. At first, how can we compute accuracy in general (without online) on predictions and targets:
```python

y_pred = ...   # tensor (N, K)  - predictions, N samples of K features
y = ...   # tensor (N, )  - targets, N samples, e.g. [0, 2, 1, 4, 5, 2, 3, 4] where K=6

pred_indices = torch.argmax(y_pred, dim=1)
num_correct = torch.eq(pred_indices, y).view(-1)  # compute the number of correctly predicted samples
num_samples = y.shape[0]

accuracy = num_correct / num_samples
```

In online manner, we have a bunch of pairs like `(y_pred, y)` and we would like to compute final accuracy. Naive way is to contatenate all `y_pred` into a single `total_y_pred` and same for `total_y` and then compute `num_correct`, `num_samples`. But this approach is not memory optimal as we have to store entire history of `(y_pred, y)`.
However, we can see that accuracy as metric can be computed as 
```
num_correct = num_correct_pair1 + num_correct_pair2 + ...num_correct = num_correct_pair1 + num_correct_pair2 + ...
accuracy = num_correct / num_samples
```

kamahori commented:
Thanks. How about this implementation?
Accumulate the features in `update` and calculate `mu`, `sigma`, FID in `compute`.

vfdev-5 reviewed (commented):
Thanks for the update ! I have some other comments.
I think it would be also good to write some basic tests.

kamahori commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,97 @@
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._value = None
```
Yes, they were not used.

kamahori commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,97 @@
+        y_pred, y = output[0].detach(), output[1].detach()
+
+        data_real, data_fake = y, y_pred
+        if isinstance(y_pred, (tuple, list)):
```
Do we not need this conditioning? (line 52 - 54)

kamahori commented:
I implemented `update` to calculate covariance in an online manner (I think this would be fine) and added a FID test.

vfdev-5 commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,95 @@
+    """
+
+    def __init__(
+        self, output_transform: Callable = lambda x: x, fid_model=None,
```
```suggestion
        self, output_transform: Callable = lambda x: x, fid_model: nn.Module = None,
```

vfdev-5 replied:
Recently, we reintroduced `device` argument to store internal vars as tensors: https://github.com/pytorch/ignite/blob/master/ignite/metrics/metric.py#L131, we have to add it here too. Please, take a look at how other metrics are implemetned, e.g. https://github.com/pytorch/ignite/blob/master/ignite/metrics/accuracy.py#L96

vfdev-5 commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,95 @@
+import os
+from typing import Callable, Sequence, Union
+
+import torch
```
```suggestion
import torch
import torch.nn as nn
```

vfdev-5 commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,95 @@
+    def reset(self) -> None:
+        self._num_of_data = 0
+
+        self._features_sum_real = None
```
Here, IMO, we can set all those things as `torch.tensor(0, device=self.device, dtype=torch.float32)` and we can simply use `+= batch_features_real` etc ...

vfdev-5 commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,95 @@
+            batch_features_real = self._fid_model(data_real)
+            batch_features_fake = self._fid_model(data_fake)
+
+        batch_features_real = batch_features_real.view(*batch_features_real.size()[:2], -1).mean(-1)
```
shouldn't it be a sum instead of mean ?

vfdev-5 reviewed (commented):
@kamahori thanks ! This looks much better now ! Now we have to write some meaningful tests...

vfdev-5 commented on tests/ignite/contrib/metrics/test_fid.py:
```diff
@@ -0,0 +1,37 @@
+    raise RuntimeError("This contrib test requires scipy to be installed.")
+
+
+def test_fid():
```
@kamahori about meaningful tests, I'd like to do the following:
- let's use a real image from internet (<300px) and add it into `tests/ignite/contrib/metrics/assets`
- manually compute FID score for this image and its slightly noised version. Otherwise use one of those referenced implementation. Let's use computed score as true score.
- compute FID score with ignite's implementation and compare scores.

kamahori commented:
I fixed the test to use an image from ImageNet (`tests/ignite/contrib/metrics/assets/fid_sample.jpg`).

vfdev-5 commented on tests/ignite/contrib/metrics/test_fid.py:
```diff
@@ -0,0 +1,46 @@
+
+    size = 10
+
+    y = torchvision.datasets.ImageFolder(
```
I think we can read the image with Pillow and use torchvision to transform it to tensor without using `datasets.ImageFolder`.
```python
from pathlib import Path

import PIL
from torchvision.transforms.functional import to_tensor


@pytest.fixture()
def img_filepath():
    fp = Path(__file__).parent / "assets" / "fid_sample.jpg"
    assert fp.exists()
    yield fp.as_posix()


def test_fid(img_filepath):
    pil_img = PIL.Image.open(img_filepath)
    y = to_tensor(pil_img)
    y_pred = y + torch.rand_like(y) * 0.01

```

kamahori commented:
I fixed the implementation, but still have some errors in CI. Do you know what is the problem? @vfdev-5 
- <https://github.com/pytorch/ignite/pull/1302/checks?check_run_id=1182940285#step:5:899>
- <https://github.com/pytorch/ignite/pull/1302/checks?check_run_id=1182940300#step:5:16>

vfdev-5 commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,89 @@
+            self._cov_real = torch.eye(len(y))
+            self._cov_fake = torch.eye(len(y_pred))
+
+    @sync_all_reduce
```
@kamahori current fail is due to bad usage of `@sync_all_reduce` which should have attributes to reduce. See https://pytorch.org/ignite/metrics.html#metrics-and-distributed-computations

vfdev-5 commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,89 @@
+    """
+
+    def __init__(
+        self, output_transform: Callable = lambda x: x, fid_model: nn.Module = None,
```
Constructor should have device argument. See Accuracy as example: https://github.com/pytorch/ignite/blob/master/ignite/metrics/accuracy.py#L135

vfdev-5 commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,89 @@
+            except ImportError:
+                raise ValueError("Argument fid_model should be set")
+
+        self._device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Device should be set with `super(FID, self).__init__(output_transform=output_transform, device=device)`

vfdev-5 commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,89 @@
+        self._features_sum_real = torch.tensor(0, device=self._device, dtype=torch.float32)
+        self._features_sum_fake = torch.tensor(0, device=self._device, dtype=torch.float32)
+
+        self._cov_real = torch.tensor(0, device=self._device, dtype=torch.float32)
```
Seems like initialization is not used as intended. Let's set them to None:
```python
self._cov_real = None
self._cov_fake = None
```

vfdev-5 commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,89 @@
+        self._features_sum_real = self._features_sum_real + batch_features_real
+        self._features_sum_fake = self._features_sum_fake + batch_features_fake
+
+        if self._num_of_data > 1:
```
Let's replace it like
```python
if self._cov_real is None:
    self._cov_real = torch.eye(len(y))
else:
     self._cov_real = self._compute_cov(batch_features_real, self._features_sum_real)
```
where we can refactor cov computation into `self._compute_cov` static function. Same for fake.

vfdev-5 commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,89 @@
+
+    @sync_all_reduce
+    def compute(self) -> Union[torch.Tensor, float]:
+        mu_real = self._features_sum_real / self._num_of_data
```
What are the shapes of `self._features_sum_real`, `self._cov_real`, `self._features_sum_fake` and `self._cov_fake` ?

kamahori replied:
When inception v3 model is used, `features_sum` is 1000 dimension vector and `cov` is 1000 * 1000 matrix.

vfdev-5 replied:
@kamahori I think this is not correct as 1000 in our case corresponds to 1000 ImageNet classes and normally it should be Inception feature before classification as here: https://github.com/clovaai/stargan-v2/blob/master/metrics/fid.py#L53

vfdev-5 commented on ignite/contrib/metrics/fid.py:
```diff
@@ -0,0 +1,89 @@
+        mu_real = self._features_sum_real / self._num_of_data
+        sigma_real = self._cov_real
+
+        mu_fake = self._features_sum_real / self._num_of_data
```
Typo with `_features_sum_real` instead of `_features_sum_fake`

vfdev-5 reviewed (commented):
@kamahori thanks for the updates. I left several comments.

kamahori commented:
I found that calculating covariance in an online manner is not efficient. So I reverted `_cov` function and implemented to compute covariance only once.

kamahori commented:
I think most parts are completed.
Do you know why these errors are happening?:

- <https://github.com/pytorch/ignite/pull/1302/checks?check_run_id=1197935105#step:5:16>
- <https://github.com/pytorch/ignite/pull/1302/checks?check_run_id=1197935271#step:5:2463>

And is there any better way to extract real values to avoid [this](https://github.com/pytorch/ignite/pull/1302/checks?check_run_id=1197935107#step:5:968)?
@vfdev-5

vfdev-5 commented:
@kamahori thanks for working on this PR which is not simple !

About XLA failing tests, let's see this a bit later as it looks like a bit unrelated to the PR.

First, I think we are incorrect about inception features. This is something to rework. Maybe, the simplest way is to replace final classification layer by identity, such that we do not need to do a lot of network surgery.

Second, I'm not that convinced about computing covariance like you changed. Now, we accumulate the sum and whole history of features... Let me also check it by myself (however, I'm a bit of lacking of time this weeks).

import-antigravity commented:
Since no one appears to have linked it, I believe this is considered to be the "reference implementation" for pytorch: https://github.com/mseitzer/pytorch-fid

sdesrozis commented:
@import-antigravity thank you for the link !

vfdev-5 commented:
closed in favor of #2049
